sui_types/src/messages.rs

```diff
@@ -311,35 +313,274 @@ where
     }
 }
 
-/// An transaction signed by a client. signature is applied on data.
+/// A transaction signed by a client, optionally signed by an authority (depending on `S`).
+/// `S` indicates the authority signing state. It can be either empty or signed.
+/// TODO: Fill comment.
 /// Any extension to Transaction should add fields to TransactionData, not Transaction.
-// TODO: this should maybe be called ClientSignedTransaction + SignedTransaction -> AuthoritySignedTransaction
-#[derive(Debug, Eq, Clone, Serialize, Deserialize)]
-pub struct Transaction {
+#[derive(Debug, Clone, Eq, Serialize, Deserialize)]
+#[serde(remote = "TransactionEnvelope")]
+pub struct TransactionEnvelope<S>
+where
+    TransactionEnvelope<S>: PartialEq,
+{
     // Deserialization sets this to "false"
     #[serde(skip)]
     pub is_checked: bool,
 
     pub data: TransactionData,
-    pub signature: Signature,
+    /// tx_signature is signed by the transaction sender, applied on `data`.
+    pub tx_signature: Signature,
+    /// auth_signature, if available, is signed by an authority, applied on `data`.
+    pub auth_signature: S,
+    // Note: If any new field is added here, make sure the Hash and PartialEq
+    // implementation are adjusted to include that new field (unless the new field
+    // does not participate in the hash and comparison).
 }
-/* TODO: check this legacy static check is not needed as an invariant
-         elsewhere. The tests pass but invariant checks without docs
-         should not be removed without a note to check.
 
-const_assert_eq!(
-    size_of::<TransactionData>() + size_of::<Signature>(),
-    size_of::<Transaction>()
-);
+impl<S: AuthoritySignInfoTrait> TransactionEnvelope<S>
```

Comment:
How about replacing `AuthoritySignInfoTrait` with its traits (which I suspect may be a bit too strong: my hunch is `Eq + Serialize + for<'de> Deserialize<'de>` would work)?